draft-irtf-cfrg-hash-to-curve.md

```diff
@@ -3764,8 +3761,9 @@ This section gives a straight-line implementation of the simplified
 SWU method for any Weierstrass curve of the form given in {{weierstrass}}.
 See {{simple-swu}} for information on the constants used in this mapping.
 
-{{sswu-opt}} gives optimized straight-line procedures that apply to specific
-classes of curves and base fields.
+This optimized, straight-line procedure applies to any classes of curves
+and base field. The constant c1 depends on a fixed, non-square value h, which
+is defined to be the smallest primitive element in the base field.
```

Comment:
@daira is `h` required to be a primitive element, or just a non-square? If the latter, could we just use the (non-square) constant `Z` that is already defined?

Otherwise, we should somewhere document how to identify the smallest primitive element, or give a pointer to an existing procedure. (Sage could work, as in the sample code, but it's not obvious to me that `F.primitive_element()` is defined to return the smallest such element, so I worry a little about things breaking in the future...)

Reply:
The algorithm that gives a square root of either s or h*s "for free" in a prime field F<sub>p</sub>, will generally require h to generate the order-2<sup>S</sup> subgroup where p-1 = 2<sup>S</sup> * T for odd T.

You could use Z = h (and I suggest that as a possibility [here](https://github.com/zcash/pasta/blob/f0f7068552a3565786cb338448cb58bc36a8314a/hashtocurve.sage#L269-L272)), but the I-D RECOMMENDS a specific Z that in general will not generate the order-T subgroup (and changing the procedure for finding the recommended Z now would not be compatible).

Reply:
Thanks for the explanation.

One more question, just to be sure that I am understanding correctly: looks like as long as the `h` value used here and in `sqrt_ratio` are the same, there's no need for different implementations to use the same `h`. Is that right? If so, is the reason to recommend the smallest such value just efficiency?

Reply:
Indeed different implementations need not use the same h. I'm not sure why the smallest value is recommended; that's not what my code used.

Reply:
> I'm not sure why the smallest value is recommended

Simply to remove choices for implementers. No other reason.

Reply:
```suggestion
is a parameter of the sqrt_ratio subroutine defined below.
```

Reply:
I'd be fine with that change as it still allows `S` to be computed deterministically.

Reply:
I said above that the Z output by `find_z_sswu` wouldn't necessarily generate the order 2<sup>S</sup> subgroup, but actually it necessarily does, as pointed out in #317.